### PR TITLE
add BaseObservationWrapper that support nested transformation

### PR DESCRIPTION
**Pull request by witwolf**

### CODE REVIEW

File: alf/environments/wrappers.py [hunk 1]
```diff
@@ -12,187 +12,166 @@
...
-from absl import logging
-
+import gin
 from tf_agents.environments import wrappers
 from tf_agents.trajectories.time_step import StepType
+from alf.utils import common
+
+
+def traverse_transform_space(space, levels, field, func):
+    """Transform space
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: Need more explanation on what this function does: given the argument `space`, `levels`, `field` and `func` what kind of output it generates?

File: alf/environments/wrappers.py [hunk 2]
```diff
@@ -12,187 +12,166 @@
...
+
+    Args:
+         space (gym.Space): space to be transformed
+         levels (list[str]): levels
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: comment what the valid values for levels

File: alf/environments/wrappers.py [hunk 3]
```diff
@@ -12,187 +12,166 @@
...
+         field (str): filed name, a multi-step path denoted by "A.B.C"
+         func (Callable): transform func
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: comment its expected arguments and outputs

File: alf/environments/wrappers.py [hunk 4]
```diff
@@ -12,187 +12,166 @@
...
+        return observation
+
+    def transform_space(self, observation_space, field=None):
+        """Transform space
+
+        Subclass should override this to perform transformation
+
+        Args:
+             observation_space (gym.Space): space to be transformed
+             field (str): field to be transformed
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: comment the format of `field`

File: alf/environments/wrappers.py [hunk 5]
```diff
@@ -12,187 +12,166 @@
...
+        return observation_space
+
+    def transform_observation(self, observation, field=None):
+        """Transform observation
+
+        Subclass should override this to perform transformation
+
+        Args:
+             observation (ndarray): observation to be transformed
+             field (str): field to be transformed
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: Returns:

File: alf/environments/wrappers.py [hunk 6]
```diff
@@ -12,187 +12,166 @@
...
+        """
+        return observation_space
+
+    def transform_observation(self, observation, field=None):
+        """Transform observation
+
+        Subclass should override this to perform transformation
+
+        Args:
+             observation (ndarray): observation to be transformed
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: The type should be ndarray or dict?

File: alf/environments/wrappers.py [hunk 7]
```diff
@@ -12,187 +12,166 @@
...
+    The key interface functions are:
+    1. transform_space(): transform space.
+    2. transform_observation(): transform observation.
+    """
+
+    def __init__(self, env, fields=None):
+        """
+        Args:
+            env (gym.Env): the gym environment
+            fields (list[str]): fields to be applied transformation
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: comment the format of fields

File: alf/environments/wrappers.py [hunk 8]
```diff
@@ -12,187 +12,166 @@
...
+        """
+        super().__init__(env)
+        paths = []
+        for field in fields or [""]:
+            # remove '' in the path
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: how does the following line perform "remove '' in the path" ?

File: alf/environments/wrappers.py (hunk 4 above)
Comment (emailweixu) [inline alf/environments/wrappers.py]: Returns:

File: alf/environments/wrappers.py (hunk 5 above)
Comment (emailweixu) [inline alf/environments/wrappers.py]: format of `field`

File: alf/environments/wrappers.py [hunk 9]
```diff
@@ -12,187 +12,166 @@
...
-            channel_order (str): one of `channels_last` or `channels_first`.
-                The ordering of the dimensions in the images.
-                `channels_last` corresponds to images with shape
-                `(height, width, channels)` (Atari's default) while
-                `channels_first` corresponds to images with shape
-                `(channels, height, width)`.
-            fields_to_stack (list of str): optional paths to the fields of the
+            channel_order (str): The ordering of the dimensions in the images.
+                should be one of `channels_last` or `channels_first`.
+            fields (list[str]): optional paths to the fields of the
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: comment the format of fields

File: alf/environments/wrappers.py [hunk 10]
```diff
@@ -12,187 +12,166 @@
...
-                spaces corresponding to the keys in FrameStack.  If is None,
-                stack all recognized spaces.
-                When input space is not a Dict, we just try to stack if we can.
-                Fields_to_stack doesn't apply anymore.
+                spaces corresponding to the keys in FrameStack.
         """
-        super().__init__(env)
-        self._frames = deque(maxlen=stack_size)
         self._channel_order = channel_order
+        if self._channel_order == 'channels_last':
```
Comment (emailweixu) [inline alf/environments/wrappers.py]: assert channel_order in [...]

File: alf/utils/common.py [hunk 11]
```diff
@@ -433,6 +433,31 @@ def _cast(obs):
     return tf.nest.map_structure(_cast, observation)
 
 
+def traverse_transform_observation(obs, levels, field, func):
+    """
```
Comment (emailweixu) [inline alf/utils/common.py]: Explain what this function does.

File: alf/utils/common.py [hunk 12]
```diff
@@ -433,6 +433,31 @@ def _cast(obs):
...
+    Args:
+        obs (nested Tensor|nested ndarray): observations to be applied the transformation
+        levels (list[str]): levels
```
Comment (emailweixu) [inline alf/utils/common.py]: explain what are the valid values of `levels` and its relationship with `field`.

File: alf/utils/common.py [hunk 13]
```diff
@@ -433,6 +433,31 @@ def _cast(obs):
...
+        field (str): full field path
```
Comment (emailweixu) [inline alf/utils/common.py]: format of field?

File: alf/utils/common.py [hunk 14]
```diff
@@ -433,6 +433,31 @@ def _cast(obs):
...
+        func (Callable): transform func
+    """
+    if not levels:
+        return func(obs, field)
+    level = levels[0]
+    if isinstance(obs, tuple) and hasattr(obs, '_fields'):
```
Comment (emailweixu) [inline alf/utils/common.py]: Perhaps add a utility function `is_namedtuple()`?

File: alf/utils/common.py [hunk 15]
```diff
@@ -449,33 +474,42 @@ def image_scale_transformer(observation, fields=None, min=-1.0, max=1.0):
...
+
+@gin.configurable
+def scale_transformer(observation, scale, dtype=tf.float32, fields=None):
+    """Scale observation
+
+    Args:
+         observation (nested Tensor): observation to be scaled
+         scale (float): scale factor
+         dtype (Dtype): The destination type.
+         fields (list[str]): fields to be scaled
```
Comment (emailweixu) [inline alf/utils/common.py]: format of `fields`

File: alf/utils/common.py [hunk 16]
```diff
@@ -803,16 +837,23 @@ def get_observation_spec():
...
     """
     assert _env, "set a global env by `set_global_env` before using the function"
     specs = _env.observation_spec()
-    return tf.nest.map_structure(
+    specs = tf.nest.map_structure(
         lambda spec: (tf.TensorSpec(spec.shape, tf.float32)
                       if spec.dtype == tf.uint8 else spec), specs)
 
+    if field:
+        for f in field.split('.'):
```
Comment (hnyu) [inline alf/utils/common.py]: namedtuple cannot be indexed by a string

File: alf/utils/common.py [hunk 17]
```diff
@@ -803,16 +837,23 @@ def get_observation_spec():
 
     https://github.com/HorizonRobotics/alf/pull/239#issuecomment-544644558
 
+    Args:
+        field (str): a multi-step path denoted by "A.B.C".
```
Comment (hnyu) [inline alf/utils/common.py]: assert field is a string

File: alf/utils/common.py [hunk 18]
```diff
@@ -433,6 +433,31 @@ def _cast(obs):
     return tf.nest.map_structure(_cast, observation)
 
 
+def traverse_transform_observation(obs, levels, field, func):
```
Comment (hnyu) [inline alf/utils/common.py]: `field` seems unnecessary here? `levels` contains the split tokens of `field`

File: alf/utils/common.py [hunk 19]
```diff
@@ -449,33 +474,42 @@ def image_scale_transformer(observation, fields=None, min=-1.0, max=1.0):
         Transfromed observation
     """
 
-    def _transform_image(obs):
+    def _transform_image(obs, field):
```
Comment (hnyu) [inline alf/utils/common.py]: `field` not used in this local function

File: alf/utils/common.py [hunk 20]
```diff
@@ -449,33 +474,42 @@ def image_scale_transformer(observation, fields=None, min=-1.0, max=1.0):
...
+    """Scale observation
+
+    Args:
+         observation (nested Tensor): observation to be scaled
+         scale (float): scale factor
+         dtype (Dtype): The destination type.
+         fields (list[str]): fields to be scaled
+    """
+
+    def _scale_obs(obs, field):
```
Comment (hnyu) [inline alf/utils/common.py]: `field` not used in this local function

File: alf/environments/wrappers.py [hunk 21]
```diff
@@ -12,187 +12,166 @@
...
+    def observation(self, observation):
+        for levels, field in self._paths:
+            observation = common.traverse_transform_observation(
+                obs=observation,
+                levels=levels,
+                field=field,
+                func=self.transform_observation)
+        return observation
+
+    def transform_space(self, observation_space, field=None):
```
Comment (hnyu) [inline alf/environments/wrappers.py]: `field` seems not necessary if this function is always called on a leaf

File: alf/environments/wrappers.py [hunk 22]
```diff
@@ -12,187 +12,166 @@
...
+
+        Subclass should override this to perform transformation
+
+        Args:
+             observation_space (gym.Space): space to be transformed
+             field (str): field to be transformed
+        """
+        return observation_space
+
+    def transform_observation(self, observation, field=None):
```
Comment (hnyu) [inline alf/environments/wrappers.py]: `field` seems not necessary if this function is always called on a leaf

File: alf/environments/wrappers.py [hunk 23]
```diff
@@ -12,187 +12,166 @@
...
-                return np.concatenate(self._frames, axis=0)
+            raise ValueError("Unsupported space:%s" % observation_space)
+
+    def transform_observation(self, observation, fields=None):
+        queue = self._frames.get(fields, None)
+        if not queue:
+            queue = deque(maxlen=self._stack_size)
+            for _ in range(self._stack_size):
+                queue.append(observation)
+            self._frames[fields] = queue
```
Comment (le-horizon) [inline alf/environments/wrappers.py]: can fields be None here?  If so, index to _frames will be None?

File: alf/environments/wrappers.py [hunk 24]
```diff
@@ -12,187 +12,166 @@
...
-            fields_to_stack (list of str): optional paths to the fields of the
+            channel_order (str): The ordering of the dimensions in the images.
+                should be one of `channels_last` or `channels_first`.
+            fields (list[str]): optional paths to the fields of the
                 Dict env.observation_space.  If specified, only use the
-                spaces corresponding to the keys in FrameStack.  If is None,
-                stack all recognized spaces.
-                When input space is not a Dict, we just try to stack if we can.
-                Fields_to_stack doesn't apply anymore.
+                spaces corresponding to the keys in FrameStack.
```
Comment (le-horizon) [inline alf/environments/wrappers.py]: If fields is None, then no field is transformed.  This is different from the current assumption of FrameStack (all fields are stacked, if fields is None).  Please add a comment about what happens if fields is None.  Also please double check to make sure current gin files have the right fields specified.  I think ac_target_nav is the only one, but I may be wrong.

File: alf/environments/wrappers.py (hunk 21 above)
Comment (witwolf) [reply alf/environments/wrappers.py]: field is necessary in some case , from `FrameStack` , field represents a unique key for a queue (many queues ,each for a field ) that handle the stacking of this field

File: alf/utils/common.py (hunk 16 above)
Comment (witwolf) [reply alf/utils/common.py]: actually , the spec will never be a namedtuple instance

see [`_spec_from_gym_space`  ](https://github.com/HorizonRobotics/alf/blob/5c648814b44e3a32e51a1f494964d14354976881/alf/environments/suite_gym.py#L65-L71),  it maps `gym.spaces.Tuple` to a tuple and we do not support unamed tuple observation in alf , and there is no space  `gym.spaces.UnamedTuple`

File: alf/environments/wrappers.py (hunk 23 above)
Comment (witwolf) [reply alf/environments/wrappers.py]: `None` can be  a key value for `dict`,  and the field will never be None here (empty str when observation is not nested )

File: alf/environments/wrappers.py (hunk 1 above)
Comment (emailweixu) [reply alf/environments/wrappers.py]: A more concrete comment: Transform the child space in `observation_space` indicated by `field` using `func`

File: alf/utils/common.py (hunk 11 above)
Comment (emailweixu) [reply alf/utils/common.py]: More detailed comment: Transform the child observation in `observation` indicated by `field` using `func`

File: alf/utils/common.py (hunk 16 above)
Comment (hnyu) [reply alf/utils/common.py]: I think the user can optionally convert a dict spec to namedtuple spec (for easier field access)? This way seems more consistent with what we have for other specs defined in ALF (they are always namedtuples).

File: alf/environments/wrappers.py (hunk 21 above)
Comment (hnyu) [reply alf/environments/wrappers.py]: Oh I see

File: alf/utils/common.py (hunk 16 above)
Comment (witwolf) [reply alf/utils/common.py]: i think it's no necessary to convert a dict spec to namedtuple,  dict is  enough easy  for field accessing in  `observation` and `observation_space`

File: alf/environments/wrappers.py (hunk 1 above)
Comment (witwolf) [reply alf/environments/wrappers.py]: updated

File: alf/environments/wrappers.py (hunk 24 above)
Comment (emailweixu) [reply alf/environments/wrappers.py]: Agree. Need to comment the behavior of "fields=None", which is to transform the non-dict observation. In this case observation cannot be dictionary.
In general, the comment of a function needs to contain enough information for the user to use the function correctly.

Comment (witwolf) [reply alf/environments/wrappers.py]: comment updated

File: alf/utils/common.py (hunk 14 above)
Comment (witwolf) [reply alf/utils/common.py]: added `is_namedtuple` in `nest_utils.py`

**Pull request merged**